sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs

```diff
@@ -282,57 +282,89 @@ private static void CloseLink(RequestResponseAmqpLink link)
+                _ = Task.Factory
+                    .FromAsync<(ReceivingAmqpLink, int, TimeSpan?, TimeSpan,
+                        TaskCompletionSource<IEnumerable<AmqpMessage>>)>
+                    (
+                        static(arguments, callback, state) =>
+                        {
+                            var (link, maxMessages, maxWaitTime, timeout, receiveMessagesCompletionSource) = arguments;
+                            return link.BeginReceiveRemoteMessages(
+                                maxMessages,
+                                TimeSpan.FromMilliseconds(20),
+                                maxWaitTime ?? timeout,
+                                callback,
+                                (link, receiveMessagesCompletionSource));
+                        },
+                        static asyncResult =>
+                        {
+                            var (link, receiveMessagesCompletionSource) =
+                                ((ReceivingAmqpLink, TaskCompletionSource<IEnumerable<AmqpMessage>>))asyncResult
+                                    .AsyncState;
+                            try
+                            {
+                                bool received =
+                                    link.EndReceiveMessages(asyncResult, out IEnumerable<AmqpMessage> amqpMessages);
+                                receiveMessagesCompletionSource.TrySetResult(received
+                                    ? amqpMessages
+                                    : Enumerable.Empty<AmqpMessage>());
+                            }
+                            catch (Exception e)
+                            {
+                                receiveMessagesCompletionSource.TrySetException(e);
+                            }
+                        },
+                        (link, maxMessages, maxWaitTime, timeout, receiveMessagesCompletionSource),
+                        default
                     ).ConfigureAwait(false);
 
-                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+                var messagesReceived = await receiveMessagesCompletionSource.Task
+                    .ConfigureAwait(false);
+
                 // If event messages were received, then package them for consumption and
                 // return them.
-
                 foreach (AmqpMessage message in messagesReceived)
                 {
                     if (_receiveMode == ServiceBusReceiveMode.ReceiveAndDelete)
                     {
                         link.DisposeDelivery(message, true, AmqpConstants.AcceptedOutcome);
                     }
+
                     receivedMessages.Add(AmqpMessageConverter.AmqpMessageToSBMessage(message));
                     message.Dispose();
                 }
 
                 return receivedMessages;
             }
+            catch (OperationCanceledException)
```

Comment:
Should we check if cancellationToken.IsCancellationRequest? And also possibly restrict to TaskCanceledException?

Reply:
IIRC, we can't restrict there because the completion source will throw the `OperationCanceledException`.   We try to normalize everything to `TaskCanceledException` around the SDK so that is what callers see.   The Service Bus and Event Hubs troubleshooting guides attribute a specific meaning to `OperationCanceled` that indicates service behavior and we wanted to avoid confusion.